src/compiler/moduleSpecifiers.ts

```diff
@@ -319,33 +319,30 @@ namespace ts.moduleSpecifiers {
             return undefined;
         }
 
-        let packageJsonContent: any | undefined;
-        const packageRootPath = moduleFileName.substring(0, parts.packageRootIndex);
+        // Simplify the full file path to something that can be resolved by Node.
+
+        let moduleSpecifier = moduleFileName;
         if (!packageNameOnly) {
-            const packageJsonPath = combinePaths(packageRootPath, "package.json");
-            packageJsonContent = host.fileExists(packageJsonPath)
-                ? JSON.parse(host.readFile(packageJsonPath)!)
-                : undefined;
-            const versionPaths = packageJsonContent && packageJsonContent.typesVersions
-                ? getPackageJsonTypesVersionsPaths(packageJsonContent.typesVersions)
-                : undefined;
-            if (versionPaths) {
-                const subModuleName = moduleFileName.slice(parts.packageRootIndex + 1);
-                const fromPaths = tryGetModuleNameFromPaths(
-                    removeFileExtension(subModuleName),
-                    removeExtensionAndIndexPostFix(subModuleName, Ending.Minimal, options),
-                    versionPaths.paths
-                );
-                if (fromPaths !== undefined) {
-                    moduleFileName = combinePaths(moduleFileName.slice(0, parts.packageRootIndex), fromPaths);
+            let packageRootIndex = parts.packageRootIndex;
+            let moduleFileNameForExtensionless: string | undefined;
+            while (true) {
+                // If the module could be imported by a directory name, use that directory's name
+                const { moduleFileToTry, packageRootPath } = tryDirectoryWithPackageJson(packageRootIndex);
+                if (packageRootPath) {
+                    moduleSpecifier = packageRootPath;
+                    break;
+                }
+                if (!moduleFileNameForExtensionless) moduleFileNameForExtensionless = moduleFileToTry;
+
+                // try with next level of directory
+                packageRootIndex = moduleFileName.indexOf(directorySeparator, packageRootIndex + 1);

```

Comment:
Just to make sure I’m understanding this correctly, the idea is:

1. we’ve located the `moduleFileName`, say, `/project/node_modules/preact/hooks/src/index.js`
2. we look at the structure of the path (as before) to call `/project/node_modules/preact` the “package root”
3. we see if `moduleFileName` is accessible (via a package.json main field) from `/project/node_modules/preact`
4. if not, go to step 3 but with one more directory appended to the end (in the first iteration of this case, `/project/node_modules/preact/hooks`)

As contrasted to the previous behavior, where we’d only try the package root before giving up and using the full `moduleFileName` (optionally with `/index.js` taken off the end).